clang/test/Analysis/pointer-sub.c

```diff
@@ -0,0 +1,74 @@
+// RUN: %clang_analyze_cc1 -analyzer-checker=alpha.core.PointerSub -verify %s
+
+void f1(void) {
+  int x, y, z[10];
+  int d = &y - &x; // expected-warning{{Subtraction of two pointers that do not point into the same array is undefined behavior}}
+  d = z - &y; // expected-warning{{Subtraction of two pointers that do not point into the same array is undefined behavior}}
+  d = &x - &x; // expected-warning{{Subtraction of two pointers that do not point into the same array is undefined behavior}}
```

Comment:
This corner case is explicitly allowed by the standard: non-array variables act as if they were single-element arrays and it's valid to do (trivial) pointer arithmetic on them.

Even calculating the past-the-end pointer of a non-array object (e.g. `&x + 1`) is a valid (but unusual) operation; however the [description on cppreference](https://en.cppreference.com/w/cpp/language/operator_arithmetic) suggests that it's UB to take the difference `(&x+1)-&x`. You should probably look this up in the standard itself if you want to be accurate.

Reply:
The tests look good, except for `&x - &x`, which should be valid.
Wrt. `&x - 1`, that should be valid as well, given that the resulting pointer is not dereferenced. However, clang's constrant interpreter disagrees with gcc's on that case: https://godbolt.org/z/dMd1rMazY
However, like I said, I believe gcc's right by accepting that code.

Reply:
> Wrt. `&x - 1`, that should be valid as well, given that the resulting pointer is not dereferenced.

The standard explicitly disallows this, see [[expr.add] part 4.2](http://eel.is/c++draft/expr.add#4.2) and 4.3 (in the most recent C++ draft standard, other versions may be different).